### PR TITLE
Async service generator

### PR DESCRIPTION
**Pull request by wildarch**
# Checklist

- [ ] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [ ] I have written tests that cover the code changes.
  - [ ] I have checked that these tests are run by
        [Cloudbuild](/cloudbuild.yaml)
  - [ ] I have updated [documentation](/docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [x] Pull request includes prototype/experimental work that is under
      construction.

Example of using the new Async service code can be found at: https://github.com/wildarch/oak/blob/async_service_with_chat/examples/chat/module/rust/src/lib.rs.

### CODE REVIEW

File: experimental/oak_async/src/grpc.rs [hunk 1]
```diff
@@ -129,3 +145,116 @@ impl<T> Drop for MultiWriter<T> {
...
+        })?
+        .as_ref()
+        // `peek` only gives us a reference to the error, and `OakError` cannot implement `Clone`,
+        // so we need to do a poor man's clone here.
+        .map_err(|e| clone_error(e))?;
+    let method_name = req1.method_name.clone();
+    Ok((method_name, requests, response_writer))
+}
+
+// Poor man's clone for `OakError`
```
Comment (wildarch) [inline experimental/oak_async/src/grpc.rs:252]: I might be able to get around this by consuming `requests` on the error path, hopefully that will allow getting rid of this ugly hack 😅

Comment (wildarch) [reply experimental/oak_async/src/grpc.rs]: That turned out to be easier than expected, sorted!

File: experimental/oak_async/src/lib.rs [hunk 2]
```diff
@@ -19,4 +19,4 @@ pub mod grpc;
 mod io;
 
 pub use executor::block_on;
-pub use io::{ChannelRead, ChannelReadStream, ReceiverAsync};
+pub use io::{run_command_loop, ChannelRead, ChannelReadStream, ReceiverAsync};
```
Comment (rbehjati) [inline experimental/oak_async/src/lib.rs:22]: Why is this needed?

File: oak_utils/src/lib.rs [hunk 3]
```diff
@@ -163,6 +163,97 @@ impl prost_build::ServiceGenerator for OakServiceGenerator {
     }
 }
 
+struct AsyncServiceGenerator;
+
+impl prost_build::ServiceGenerator for AsyncServiceGenerator {
```
Comment (rbehjati) [inline oak_utils/src/lib.rs:168]: Is it possible to add a simple example that uses this? It would be nice to see the output of the generator.

Comment (wildarch) [reply oak_utils/src/lib.rs:168]: I have a separate branch that modifies the chat example to use the async service generator: https://github.com/wildarch/oak/blob/async_service_with_chat/examples/chat/module/rust/src/lib.rs.

I would prefer not to add it to this PR yet, to keep the changes small and self-contained.

File: experimental/oak_async/src/lib.rs (hunk 2 above)
Comment (wildarch) [reply experimental/oak_async/src/lib.rs:22]: The `io` module is not public, so we need to manually export the function here.

File: experimental/oak_async/src/grpc.rs [hunk 4]
```diff
@@ -129,3 +145,116 @@ impl<T> Drop for MultiWriter<T> {
...
+}
+
+// Unpin is needed to be able to peek at the first request before passing it
+// on to the handler.
+impl<T: Decodable> Unpin for GrpcRequestStream<T> {}
+
+// This should only be used by the Oak service generator. Not a stable API.
+#[doc(hidden)]
+impl GrpcRequestStream<()> {
+    pub fn new(inner: ChannelReadStream<GrpcRequest>) -> GrpcRequestStream<()> {
```
Comment (rbehjati) [inline experimental/oak_async/src/grpc.rs:165]: Can this be non-public?

File: experimental/oak_async/src/grpc.rs [hunk 5]
```diff
@@ -129,3 +145,116 @@ impl<T> Drop for MultiWriter<T> {
...
+                    };
+                    T::decode(&message)
+                })
+            })
+        })
+    }
+}
+
+// This should only be used by the Oak service generator. Not a stable API.
+#[doc(hidden)]
```
Comment (rbehjati) [inline experimental/oak_async/src/grpc.rs:221]: Adding a new struct, with specific field names, for the return type might be overkill, but I think it is still useful to have a bit of documentation here, explaining what this function produces.

File: experimental/oak_async/src/grpc.rs [hunk 6]
```diff
@@ -129,3 +145,116 @@ impl<T> Drop for MultiWriter<T> {
...
+pub async fn invocation_to_requests_and_writer(
```
Comment (rbehjati) [inline experimental/oak_async/src/grpc.rs:222]: > 
```suggestion
pub async fn invocation_to_requests_and_response_writer(
```

File: experimental/oak_async/src/grpc.rs [hunk 7]
```diff
@@ -129,3 +145,116 @@ impl<T> Drop for MultiWriter<T> {
...
+    })?;
+    let mut requests = GrpcRequestStream::new(receiver.receive_stream());
+    let requests_pinned = core::pin::Pin::new(&mut requests);
+    let sender = invocation.sender.ok_or_else(|| {
+        error!("Invocation did not have a sender");
+        OakError::OakStatus(OakStatus::ErrInvalidArgs)
+    })?;
+    let response_writer = ChannelResponseWriter::new(sender);
+
+    let req1 = match requests_pinned
```
Comment (rbehjati) [inline experimental/oak_async/src/grpc.rs:237]: `req1` is only used for getting the `method_name` from it. I think it is more readable to merge this and the next statement, and only declare `method_name`.

File: experimental/oak_async/src/io.rs [hunk 8]
```diff
@@ -128,10 +128,31 @@ pub trait ReceiverAsync {
...
+    T: Decodable + Send,
+    F: FnOnce(ChannelReadStream<T>) -> R,
+    R: Future<Output = ()> + 'static,
+{
+    match crate::block_on(handler(receiver.receive_stream())) {
+        Ok(()) => {}
+        Err(OakStatus::ErrTerminated) => {
+            info!("Received termination status, terminating command loop");
+        }
+        Err(e) => panic!("Command loop received non-termination error: {:?}", e),
```
Comment (rbehjati) [inline experimental/oak_async/src/io.rs:156]: In the other `run_command_loop` in the SDK, we log an error and return in case of an error. Does it make sense to do the same thing here as well instead of using `panic!`?

File: oak_utils/src/lib.rs (hunk 3 above)
Comment (rbehjati) [reply oak_utils/src/lib.rs:168]: Thanks for the link :)

**Review by rbehjati — APPROVED** [review]
LGTM. Thanks!

File: experimental/oak_async/src/grpc.rs (hunk 4 above)
Comment (wildarch) [reply experimental/oak_async/src/grpc.rs:165]: Aah yes, this can actually by private now. Thanks for spotting!

File: experimental/oak_async/src/io.rs (hunk 8 above)
Comment (wildarch) [reply experimental/oak_async/src/io.rs:156]: I think it makes sense to have the same behaviour for both, will change to `error!` instead. Thanks!

File: experimental/oak_async/src/grpc.rs (hunk 5 above)
Comment (wildarch) [reply experimental/oak_async/src/grpc.rs:221]: I think it's a good idea, done!
